internal/uvm/vpmem_mapped.go

```diff
@@ -0,0 +1,307 @@
+	MaxMappedDeviceCount    = 128
+)
+
+type mappedDeviceInfo struct {
+	vPMemInfoDefault
+	deviceOffset uint64
+	deviceSize   uint64
+}
+
+type vPMemInfoMulti struct {
+	memory.PoolAllocator
+	maxSize              uint64
+	maxMappedDeviceCount uint32
+	mappings             map[string]*mappedDeviceInfo
+}
+
+func newVPMemMappedDevice(hostPath, uvmPath string, deviceOffset, deviceSize uint64) *mappedDeviceInfo {
+	return &mappedDeviceInfo{
+		vPMemInfoDefault: vPMemInfoDefault{
+			hostPath: hostPath,
+			uvmPath:  uvmPath,
+			refCount: 1,
+		},
+		deviceOffset: deviceOffset,
+		deviceSize:   deviceSize,
+	}
+}
+
+func newPackedVPMemDevice() *vPMemInfoMulti {
+	return &vPMemInfoMulti{
+		PoolAllocator:        memory.NewPoolMemoryAllocator(),
+		maxSize:              DefaultVPMemSizeBytes,
+		mappings:             make(map[string]*mappedDeviceInfo),
+		maxMappedDeviceCount: MaxMappedDeviceCount,
+	}
+}
+
+func pageAlign(t uint64) uint64 {
+	if t%PageSize == 0 {
+		return t
+	}
+	return (t/PageSize + 1) * PageSize
+}
+
+// newMappedVPMemModifyRequest creates an hcsschema.ModifySettingsRequest to modify VPMem devices/mappings
+// for the multi-mapping setup
+func newMappedVPMemModifyRequest(ctx context.Context, rType string, deviceNumber uint32, md *mappedDeviceInfo, uvm *UtilityVM) (*hcsschema.ModifySettingRequest, error) {
+	request := &hcsschema.ModifySettingRequest{
+		RequestType: rType,
+	}
+	pmem := uvm.vpmemDevicesPacked[deviceNumber]
+	switch rType {
+	case requesttype.Add:
+		if pmem == nil {
+			request.Settings = hcsschema.VirtualPMemDevice{
+				ReadOnly:    true,
+				HostPath:    md.hostPath,
+				ImageFormat: "Vhd1",
+			}
+			request.ResourcePath = fmt.Sprintf(vPMemControllerResourceFormat, deviceNumber)
+		} else {
+			request.Settings = hcsschema.VirtualPMemMapping{
+				HostPath:    md.hostPath,
+				ImageFormat: "Vhd1",
+			}
+			request.ResourcePath = fmt.Sprintf(vPMemDeviceResourceFormat, deviceNumber, md.deviceOffset)
+		}
+	case requesttype.Remove:
+		if pmem == nil {
+			return nil, errors.Errorf("no device found at location %d", deviceNumber)
+		}
+		if len(pmem.mappings) == 1 {
+			request.ResourcePath = fmt.Sprintf(vPMemControllerResourceFormat, deviceNumber)
+		} else {
+			request.ResourcePath = fmt.Sprintf(vPMemDeviceResourceFormat, deviceNumber, md.deviceOffset)
+		}
+	default:
+		return nil, errors.New("unsupported request type")
+	}
+
+	request.GuestRequest = guestrequest.GuestRequest{
+		ResourceType: guestrequest.ResourceTypeVPMemDevice,
+		RequestType:  rType,
+		Settings: guestrequest.LCOWMappedVPMemDevice{
+			DeviceNumber: deviceNumber,
+			MountPath:    md.uvmPath,
+			MappingInfo: &guestrequest.LCOWMappedLayer{
+				DeviceOffsetInBytes: md.deviceOffset,
+				DeviceSizeInBytes:   md.deviceSize,
+			},
+		},
+	}
+	log.G(ctx).WithFields(logrus.Fields{
+		"deviceNumber": deviceNumber,
+		"hostPath":     md.hostPath,
+		"uvmPath":      md.uvmPath,
+	}).Debugf("new mapped VPMem modify request: %s", request)
```

Comment:
Will the struct get printed out correctly with %s and not %v/%+v? Might be better to just make a couple more fields in `WithFields` with what we care about logging and just get rid of the formatting and swap to .`Debug`